File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(begin(nums),end(nums));
        int n=nums.size();
        vector<vector<int>> res; 
        for(int i=0;i<n-3;i++){
            while(i>0 && i<n-3 && nums[i]==nums[i-1]) i++;
            for(int j=i+1;j<n-2;j++){
                while(j>i+1 && j<n-2 && nums[j]==nums[j-1]) j++;
                int k=j+1,l=n-1;
                while(k<l){
                    long long  sum = (long long)nums[i] +
                                    (long long)nums[j] +
                                    (long long)nums[k] +
                                    (long long)nums[l];
                    if(sum==target){
                        res.push_back({nums[i],nums[j],nums[k++],
                                    nums[l--]});
                        while(k<l && nums[k]==nums[k-1]) k++;
                        while(k<l && nums[l]==nums[l+1]) l--;
                    }
                    else if(sum>target) l--;
                    else k++;
                }
            }
        }
        return res;
    }
};

File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums[:] = sorted(nums) 
        n = len(nums)
        res=[]
        for i in range(len(nums)): 
            if i>0 and nums[i]==nums[i-1]: 
                continue 
            for j in range(i+1,len(nums)): 
                if j>i+1 and nums[j]==nums[j-1]:
                    continue 
                l,r=j+1,n-1
                while l<r: 
                    total = nums[i]+nums[j]+nums[l]+nums[r] 
                    if total>target: 
                        r-=1
                    elif total<target: 
                        l+=1 
                    else: 
                        res.append([nums[i],nums[j],nums[l],nums[r]])
                        l+=1
                        r-=1
                        while l<r and l>j+1 and nums[l]==nums[l-1]:
                            l+=1 
                        while l<r and r<n-1 and nums[r]==nums[r+1]:
                            r-=1
        
        return res
